lib/sdf-server/src/server/service/graphviz.rs

```diff
@@ -246,10 +283,99 @@ pub async fn schema_variant(
     Ok(Json(GraphVizResponse {
         nodes,
         edges,
-        root_node_id,
+        root_node_id: Some(root_node_id),
     }))
 }
 
+pub async fn components(
+    HandlerContext(builder): HandlerContext,
+    AccessBuilder(request_ctx): AccessBuilder,
+    Query(request): Query<GraphVizRequest>,
+) -> GraphVizResult<Json<GraphVizResponse>> {
+    let ctx = builder.build(request_ctx.build(request.visibility)).await?;
+
+    let workspace_snapshot = ctx.workspace_snapshot()?;
+
+    let mut func_nodes = vec![];
+    let mut nodes = vec![];
+    let mut edges = vec![];
+    let mut added_nodes = HashSet::new();
+    let mut added_edges = HashSet::new();
+
+    let components = Component::list(&ctx).await?;
+
+    for component in &components {
+        if !added_nodes.contains(&component.id()) {
+            added_nodes.insert(component.id());
+        } else {
+            continue;
+        }
+
+        let node_weight = workspace_snapshot.get_node_weight_by_id(component.id()).await?;
+        let node  = GraphVizNode {
+            id: node_weight.id(),
+            content_kind: node_weight.content_address_discriminants(),
+            node_kind: node_weight.into(),
+            name: Some(component.name(&ctx).await?.to_owned()),
+        };
+        nodes.push(node);
+
+
+        let mut work_queue: VecDeque<Ulid> = VecDeque::from([component.id().into()]);
+        while let Some(id) = work_queue.pop_front() {
+            for target in workspace_snapshot.all_outgoing_targets(id).await? {
+                work_queue.push_back(target.id());
+                if !added_edges.contains(&(id, target.id())) {
+                    added_edges.insert((id, target.id()));
+
+                    let rel_edges = workspace_snapshot
+                    .get_edges_between_nodes(id, target.id())
+                    .await?;
+                    rel_edges.iter().enumerate().for_each(|(_, edge)| {
+                        edges.push(GraphVizEdge {
+                            from: id,
+                            to: target.id(),
+                            edge_weight_kind: edge.kind().into(),
+                            direction: Direction::Outgoing,
+                        });
+                    });
+                }
```

Comment:
Hmm... this is only dealing with `Components` right now, but I'm wondering if you could make it generic for _everything_ in the graph. It would likely be a simpler algorithm too (e.g. go through every single edge in the graph and send them up as views).

Reply:
Do you mean like the `node_edges` endpoint?